hips/hip-0015.md

```diff
@@ -0,0 +1,126 @@
+---
+hip: "0015"
+title: "Annotation for Images used by Helm Charts"
+authors:
+  [
+    "Tom Runyon <tom@defenseunicorns.com>",
+    "Scott Rigby <scott@weave.works>",
+    "Andrew Block <ablock@redhat.com>",
+  ]
+created: "2021-10-07"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+The images used by a Helm chart is not easily discoverable in the chart definition. When deploying Operators and CRDs, the images needed to instantate all the pods for a Helm release may not be present by templating the default values for a Helm chart, whic provides blockers when automatically discovering dependencies for Bill of Materials.
+
+## Motivation
+
+Without a standard methodology for defining the images used by a chart, third party applications that require metadata for Helm Charts are using custom keys unique to their application, e.g. [Artifact Hub](https://artifacthub.io/docs/topics/annotations/helm/#supported-annotations) or [Openshift](https://docs.openshift.com/container-platform/4.8/operators/operator_sdk/osdk-generating-csvs.html#olm-enabling-operator-for-restricted-network_osdk-generating-csvs). By Standardizing metadata, charts can be built to be consumable by a varity of tools for metadata, bill of materials, airgap deployments, etc. Current solutions of dynamically discovering the images needed by templating the chart do not account for all possible templating paths for the chart, nor capture images that are deployed as part of an Operator paradigm and not defined explicitly as pods in the helm chart.
+
+## Rationale
+
+The `helm.sh` prefix identifies the standard belonging to the Helm community, and the `/images` path clearly identifies the type of objects being identified. The annotation will define a string that can be parsed into a list that the chart maintainer will manage as they knowingly adjust the images needed for the chart's deployment. Each element of this image list will contain at least a `name` for the image as well as a location for the image in the `image` field. and an optional `dependency` field described in the specification. The elements _may_ have additional properties to meet additional use cases, such as the `whitelisted` field used by ArtifactHub, or providing a public key for signed images. The additional fields are not part of this standard, but how they get attached to each image _is_.
+
+## Specification
+
+A helm chart would be define the images in the `annotations` field:
+
+```yaml
+apiVersion: v1
+version: 6.0.0
+appVersion: 6.0.0
+name: podinfo
+engine: gotpl
+description: Podinfo Helm chart for Kubernetes
+home: https://github.com/stefanprodan/podinfo
+maintainers:
+  - email: stefanprodan@users.noreply.github.com
+    name: stefanprodan
+sources:
+  - https://github.com/stefanprodan/podinfo
+kubeVersion: ">=1.19.0-0"
+annotations:
```

Comment:
`images` as a new top level element in the `Chart.yaml` file cannot happen until the next major version of Helm is introduced. This is due to the use of strict handling when the `Chart.yaml` file is loaded. This is done for security purposes and was a finding in a security audit. You can refer to the audits for more details on the issue.

An annotation is backwards compatible and lets us do it now.

Reply:
What's the right way to capture this top level image field as part of the 4.X release so that can be tracked?

Reply:
Two possibilities: Open a new HIP describing the change or mark this HIP as Helm 4, and not implement it in Helm 3. In theory, Helm 4's feature list is supposed to be built entirely from HIPs so that we have a documented forward path. This is after receiving backlash from people who felt that Helm 3 did not have an adequate set of planning documents.